subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -0,0 +1,362 @@
+ * ```
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "gstframerate.h"
+
+GST_DEBUG_CATEGORY_STATIC (gst_framerate_debug);
+#define GST_CAT_DEFAULT gst_framerate_debug
+
+#define _do_init \
+    GST_DEBUG_CATEGORY_INIT (gst_framerate_debug, "framerate", 0, "framerate tracer");
+#define gst_framerate_tracer_parent_class parent_class
+G_DEFINE_TYPE_WITH_CODE (GstFramerateTracer, gst_framerate_tracer,
+    GST_TYPE_TRACER, _do_init);
+
+#define DEFAULT_FPS_UPDATE_INTERVAL_MS 500      /* 500 ms */
+
+static void framerate_query_stack_destroy (gpointer data);
+
+static GQuark framerate_probe_id;
+static GQuark sub_framerate_probe_id;
+static GQuark drop_sub_framerate_quark;
+
+static GstTracerRecord *tr_framerate;
+static GstTracerRecord *tr_element_framerate;
+static GstTracerRecord *tr_element_reported_framerate;
+
+/* The private stack for each thread */
+static GPrivate framerate_query_stack =
+G_PRIVATE_INIT (framerate_query_stack_destroy);
+
+struct FramerateQueryTableValue
+{
+  GstElement *peer_element;
+  guint64 min;
+  guint64 max;
+};
+
+/* data helpers */
+
+/*
+ * Get the element/bin owning the pad.
+ *
+ * in: a normal pad
+ * out: the element
+ *
+ * in: a proxy pad
+ * out: the element that contains the peer of the proxy
+ *
+ * in: a ghost pad
+ * out: the bin owning the ghostpad
+ */
+/* TODO(ensonic): gst_pad_get_parent_element() would not work here, should we
+ * add this as new api, e.g. gst_pad_find_parent_element();
+ */
+static GstElement *
+get_real_pad_parent (GstPad *pad)
+{
+  GstObject *parent;
+
+  if (!pad)
+    return NULL;
+
+  parent = gst_object_get_parent (GST_OBJECT_CAST (pad));
+
+  /* if parent of pad is a ghost-pad, then pad is a proxy_pad */
+  if (parent && GST_IS_GHOST_PAD (parent)) {
+    GstObject *tmp;
+    pad = GST_PAD_CAST (parent);
+    tmp = gst_object_get_parent (GST_OBJECT_CAST (pad));
+    gst_object_unref (parent);
+    parent = tmp;
+  }
+  return GST_ELEMENT_CAST (parent);
+}
+
+/* hooks */
+
+
+static gboolean
+gst_framerate_tracer_display_current_fps (GstFramerateTracer *self,
+    GstClockTime ts)
+{
+  guint64 frames_rendered, frames_dropped;
+  gdouble rr, dr, average_fps;
+  gchar fps_message[256];
+  gdouble time_diff, time_elapsed;
+  GstClockTime current_ts = gst_util_get_timestamp ();
+
+  frames_rendered = g_atomic_int_get (&self->frames_rendered);
+  frames_dropped = g_atomic_int_get (&self->frames_dropped);
+
+  if ((frames_rendered + frames_dropped) == 0) {
+    /* in case timer fired and we didn't yet get any QOS events */
+    return TRUE;
+  }
+
+  time_diff = (gdouble) (current_ts - self->last_ts) / GST_SECOND;
+  time_elapsed = (gdouble) (current_ts - self->start_ts) / GST_SECOND;
+
+  rr = (gdouble) (frames_rendered - self->last_frames_rendered) / time_diff;
+  dr = (gdouble) (frames_dropped - self->last_frames_dropped) / time_diff;
+
+  average_fps = (gdouble) frames_rendered / time_elapsed;
+
+  if (self->max_fps == -1 || rr > self->max_fps) {
```

Comment:
`rr` is in the range `0..G_MAXDOUBLE`, that means it is _always_ higher then `-1`. So no matter what value it has, if `rr > self->max_fps` is enough, since for `self->max_fps == -1` this one will always be `true`.

Reply:
I don't understand. Should I delete this?

```
  if (self->max_fps == -1 || rr > self->max_fps) {
    self->max_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated max-fps to %f", rr);
  }
  if (self->min_fps == -1 || rr < self->min_fps) {
    self->min_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated min-fps to %f", rr);
  }
```

Reply:
IMHO it should be
```c
  if (rr > self->max_fps) {
    self->max_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated max-fps to %f", rr);
  }
  if (rr < self->min_fps) {
    self->min_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated min-fps to %f", rr);
  }
```